src/dataframe/dataframe.jl

```diff
@@ -368,7 +380,7 @@ end
 function Base.setindex!(df::DataFrame,
                 v::Any,
                 col_ind::ColumnIndex)
-    insert_single_column!(df, upgrade_scalar(df, v), col_ind)
+    fill_single_column!(df, v, col_ind)
```

Comment:
It looks like https://github.com/JuliaLang/julia/pull/16310 has closed the performance difference, although that means it still exists on v0.4. I think keeping `fill!` might be better for now.